text/0000-dockerfiles.md

```diff
@@ -0,0 +1,113 @@
+# Meta
+[meta]: #meta
+- Name: Support Dockerfiles
+- Start Date: 2021-06-30
+- Author(s): sclevine
+- RFC Pull Request: (leave blank)
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: [RFC0069](https://github.com/buildpacks/rfcs/blob/main/text/0069-stack-buildpacks.md), [RFC#167](https://github.com/buildpacks/rfcs/pull/167)
+- Depends on: [RFC#172](https://github.com/buildpacks/rfcs/pull/172)
+
+# Summary
+[summary]: #summary
+
+This RFC introduces functionality for customizing base images, as an alternative to stackpacks.
+
+# Motivation
+[motivation]: #motivation
+
+Relying on Dockerfiles for base image generation and manipulation allows us to apply buildpacks only to the problem that they solve best: managing application runtimes and dependencies.
+
+# What it is
+[what-it-is]: #what-it-is
+
+This RFC proposes that we replace stackpacks with multi-purpose build-time and runtime Dockerfiles.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+Note: kaniko, BuildKit, and/or the original Docker daemon may be used to apply Dockerfiles at the platform's discretion. 
+
+### App-specified Dockerfiles
+
+A buildpack app may have a build.Dockerfile and/or run.Dockerfile in its app directory. A run.Dockerfile is applied to the selected runtime base image after the detection phase. A build.Dockerfile is applied to the build-time base image before the detection phase.
+
+Both Dockerfiles must accept `base_image` and `build_id` args. The `base_image` arg allows the lifecycle to specify the original base image. The `build_id` arg allows the app developer to bust the cache after a certain layer and must be defaulted to `0`.  
+
+A runtime base image may indicate that it preserves ABI compatibility by adding the label `io.buildpacks.rebasable=true`. In the case of app-specified Dockerfiles, `io.buildpacks.rebasable=false` is set automatically before `run.Dockerfile` is applied and must be explicitly set to `true` if desired. Rebasing an app without this label set to `true` requires passing a new `--force` flag to `pack rebase`.
```

Comment:
I assume this will be tue for`/cnb/lifecycle/rebaser` also?

What is the behavior when trying to rebase an image that has `io.buildpacks.rebasable=false`? Error or no-op?

Reply:
> I assume this will be true for/cnb/lifecycle/rebaser also?

Definitely

> What is the behavior when trying to rebase an image that has io.buildpacks.rebasable=false? Error or no-op?

Error -- the user should know that the rebase did not succeed, so they can rebuild.

Reply:
I wonder about the concept of being "rebasable" and if users will understand that. I could see someone just marking rebasable = true because it sounds better or having users just always add the `-f` flag (cause what could go wrong with that?), and not really understand what it means to be rebasable or ABI compatible. 

It makes me wonder what will happen if you force it to true & try to rebase? Are you going to get clear and obvious errors or are you going to get very subtle and difficult to debug errors? Best case, maybe you get shared library loading problems as the app starts up and you know something is horribly wrong? Maybe worst case, things fail later as your code tries to do certain things and load additional libraries (hopefully soon enough after the rebase that you can connect the dots)? 

I'm not sure there's much you can do about it. Maybe you could have a safety check like force the user to confirm when they have a Dockerfile that contains rebasable=true? That still wouldn't solve it though. 

Maybe we just need a note to flag the importance of documenting this label so users know when it's safe to mark it as true.

Reply:
> I wonder about the concept of being "rebasable" and if users will understand that

Such a property `io.buildpacks.rebasable=true` will certainly create confusion as the end users will not understand the reason why the image is not rebasable (Is it because, for licensing , security, business or technical reasons it cannot be rebased ?). What should they do then when the `pack rebase` will report such a message `image not rebasable` (rebuild, contact their `builder`s team, security advisor, ops team,...) ?

Reply:
Aside from clear error messages and warnings, I think the alternative would be to permanently disable rebase any time a Dockerfile is used to extend an existing stack. I prefer clear error messages and warnings.

Reply:
My $0.02, we should optimize for the common use case. We know that most users will want to use this feature add some additional libraries to the stack. The common request is "I need libfoo or foo-cli to compile/run my app, how do I install that?". Since that's likely what most users will do, we should optimize for that case. If it's generally going to produce a rebasable container, then it seems reasonable to leave it up to the user (who, I believe, will very likely mark it true without a lot of thought). If it's not generally going to produce a rebasable container, we should just always mark the container as not rebasable permanently. If it's complicated enough that we can't clearly tell if the happy path will produce rebasable containers, then we should probably mark an extended container as permanently not rebasable.

Reply:
> who, I believe, will very likely mark it true without a lot of thought

As specified, it would be reset to false before any subsequent Dockerfile is applied, and would not need to be present in examples. The user would have to look up the label in the docs and intentionally add it to enable rebasing.

Reply:
Yes, I think that's exactly what'll happen. StackOverflow, copy & paste, boom. They set it to rebasable without understanding anything (or at best very little) about what that means. For what it's worth though, I think the same will happen with the rebase command, users will add `--force` because it "works" not really understanding the nuance.

I am a pessimist regarding stuff like this, 10+ years in support will do that to you though :) I'm OK to resolve this comment if you'd prefer to handle this through docs/errors & warnings. You either allow folks to shoot themselves in the foot and potentially mark something rebasable that's not, or it's never rebasable when you use app Dockerfiles. I can get why you want to allow it for responsible users, rebasing is a nice feature.

Reply:
I think the new hooks model partially mitigates this by making the Dockerfiles a builder-level concern.

Reply:
(closing comments so that PR page loading stops timing out)